trl/trainer/ppo_trainer.py

```diff
@@ -982,6 +982,18 @@ def compute_rewards(
             rewards.append(reward)
         return torch.stack(rewards), torch.stack(non_score_rewards)
 
+    def _kl_penalty(self, logprob: torch.FloatTensor, ref_logprob: torch.FloatTensor) -> torch.FloatTensor:
+        if self.config.kl_penalty == "kl":
+            return logprob - ref_logprob
+
+        if self.config.kl_penalty == "abs":
+            return (logprob - ref_logprob).abs()
+
+        if self.config.kl_penalty == "mse":
+            return 0.5 * (logprob - ref_logprob).square()
+
+        raise NotImplementedError
+
```

Comment:
Ohhh this change looks interesting!! Maybe you could run the experiments for 10 random seeds to account for randomness, especially the negative KL divergence issue does not always occur?

You can run 

```
WANDB_TAGS="kl-penalty-kl" python benchmark/benchmark.py \
    --command "python examples/sentiment/scripts/gpt2-sentiment.py --log_with wandb" \
    --num-seeds 10 \
    --start-seed 1 \
    --workers 10 \
    --slurm-nodes 1 \
    --slurm-gpus-per-task 1 \
    --slurm-ntasks 1 \
    --slurm-total-cpus 12 \
    --slurm-template-path benchmark/trl.slurm_template
WANDB_TAGS="kl-penalty-mse" python benchmark/benchmark.py \
    --command "python examples/sentiment/scripts/gpt2-sentiment.py --log_with wandb --kl_penalty mse" \
    --num-seeds 10 \
    --start-seed 1 \
    --workers 10 \
    --slurm-nodes 1 \
    --slurm-gpus-per-task 1 \
    --slurm-ntasks 1 \
    --slurm-total-cpus 12 \
    --slurm-template-path benchmark/trl.slurm_template
WANDB_TAGS="kl-penalty-abs" python benchmark/benchmark.py \
    --command "python examples/sentiment/scripts/gpt2-sentiment.py --log_with wandb --kl_penalty abs" \
    --num-seeds 10 \
    --start-seed 1 \
    --workers 10 \
    --slurm-nodes 1 \
    --slurm-gpus-per-task 1 \
    --slurm-ntasks 1 \
    --slurm-total-cpus 12 \
    --slurm-template-path benchmark/trl.slurm_template
```

to run the benchmark experiments for 10 random seeds for each of the three options. The you can prob run the following to generate plots and a wandb report to ease the process :) 


```
pip install openrlbenchmark==0.2.1a4
python -m openrlbenchmark.rlops_multi_metrics \
    --filters '?we=edbeeching&wpn=trl&xaxis=_step&ceik=trl_ppo_trainer_config.value.tracker_project_name&cen=trl_ppo_trainer_config.value.log_with&metrics=env/reward_mean&metrics=objective/kl' \
        'wandb?tag=kl-penalty-kl&cl=kl-penalty-kl (sentiment analysis)' \
        'wandb?tag=kl-penalty-mse&cl=kl-penalty-mse (sentiment analysis)' \
        'wandb?tag=kl-penalty-abs&cl=kl-penalty-abs (sentiment analysis)' \
    --env-ids trl \
    --check-empty-runs \
    --pc.ncols 2 \
    --pc.ncols-legend 1 \
    --output-filename static/0compare \
    --scan-history --report
```

Reply:
I run these benchmarks but have an error collating the results, perhaps related to a couple of crashed runs?



(trl_dev) edward@ip-10-52-17-32:~/work/trl_dev/trl$ python -m openrlbenchmark.rlops_multi_metrics \
>     --filters '?we=edbeeching&wpn=trl&xaxis=_step&ceik=trl_ppo_trainer_config.value.tracker_project_name&cen=trl_ppo_trainer_config.value.log_with&metrics=env/reward_mean&metrics=objective/kl' \
>         'wandb?tag=kl-penalty-kl&cl=kl-penalty-kl (sentiment analysis)' \
>         'wandb?tag=kl-penalty-mse&cl=kl-penalty-mse (sentiment analysis)' \
>         'wandb?tag=kl-penalty-abs&cl=kl-penalty-abs (sentiment analysis)' \
>     --env-ids trl \
>     --check-empty-runs \
>     --pc.ncols 2 \
>     --pc.ncols-legend 1 \
>     --output-filename static/0compare \
>     --scan-history --report
wandb: Thanks for trying out the Report API!
wandb: For a tutorial, check out https://colab.research.google.com/drive/1CzyJx1nuOS4pdkXa2XPaRQyZdmFmLmXV
wandb: 
wandb: Try out tab completion to see what's available.
wandb:   ∟ everything:    `wr.<tab>`
wandb:       ∟ panels:    `wr.panels.<tab>`
wandb:       ∟ blocks:    `wr.blocks.<tab>`
wandb:       ∟ helpers:   `wr.helpers.<tab>`
wandb:       ∟ templates: `wr.templates.<tab>`
wandb:       
wandb: For bugs/feature requests, please create an issue on github: https://github.com/wandb/wandb/issues
{
│   'wandb_project_name': 'trl',
│   'wandb_entity': 'edbeeching',
│   'custom_env_id_key': 'trl_ppo_trainer_config.value.tracker_project_name',
│   'custom_exp_name_key': 'trl_ppo_trainer_config.value.log_with',
│   'custom_xaxis_key': '_step',
│   'metrics': [
│   │   'env/reward_mean',
│   │   'objective/kl'
│   ]
}
/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/openrlbenchmark/dataset/edbeeching/trl/offline.sqlite
========= wandb?tag=kl-penalty-kl&cl=kl-penalty-kl (sentiment analysis)
wandb ({'tag': ['kl-penalty-kl'], 'cl': ['kl-penalty-kl (sentiment analysis)']}) in trl has 10 runs
┣━━ lucky-plant-521 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ effortless-thunder-515 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ pious-plasma-518 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ swept-jazz-515 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ absurd-dragon-515 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ comic-cosmos-519 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ vocal-wave-520 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ revived-night-514 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ grateful-puddle-495 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
┣━━ fancy-energy-496 with tags = ['kl-penalty-kl', 'v0.4.6-11-g2823e3d']
{'$and': [{'config.trl_ppo_trainer_config.value.tracker_project_name': 'trl'}, {'tags': {'$in': ['kl-penalty-kl']}}, {'config.trl_ppo_trainer_config.value.log_with': 'wandb'}]}
========= wandb?tag=kl-penalty-mse&cl=kl-penalty-mse (sentiment analysis)
wandb ({'tag': ['kl-penalty-mse'], 'cl': ['kl-penalty-mse (sentiment analysis)']}) in trl has 10 runs
┣━━ expert-waterfall-513 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ dulcet-flower-510 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ balmy-music-509 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ rosy-darkness-504 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ chocolate-shape-505 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ effortless-capybara-502 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ fragrant-plant-499 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ skilled-cloud-500 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ stilted-donkey-498 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
┣━━ daily-silence-493 with tags = ['kl-penalty-mse', 'v0.4.6-11-g2823e3d']
{'$and': [{'config.trl_ppo_trainer_config.value.tracker_project_name': 'trl'}, {'tags': {'$in': ['kl-penalty-mse']}}, {'config.trl_ppo_trainer_config.value.log_with': 'wandb'}]}
========= wandb?tag=kl-penalty-abs&cl=kl-penalty-abs (sentiment analysis)
wandb ({'tag': ['kl-penalty-abs'], 'cl': ['kl-penalty-abs (sentiment analysis)']}) in trl has 10 runs
┣━━ ethereal-sun-511 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ radiant-dew-511 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ lemon-dawn-508 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ misty-tree-507 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ deft-water-506 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ fearless-sea-503 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ summer-universe-501 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ laced-cherry-494 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ zany-blaze-492 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
┣━━ divine-music-497 with tags = ['kl-penalty-abs', 'v0.4.6-11-g2823e3d']
{'$and': [{'config.trl_ppo_trainer_config.value.tracker_project_name': 'trl'}, {'tags': {'$in': ['kl-penalty-abs']}}, {'config.trl_ppo_trainer_config.value.log_with': 'wandb'}]}
collecting runs for trl
loading <Run edbeeching/trl/y8txzxqx (finished)> https://wandb.ai/edbeeching/trl/runs/y8txzxqx
loading <Run edbeeching/trl/tg9xheei (finished)> https://wandb.ai/edbeeching/trl/runs/tg9xheei
loading <Run edbeeching/trl/r0x7af7q (finished)> https://wandb.ai/edbeeching/trl/runs/r0x7af7q
loading <Run edbeeching/trl/eci6haqe (finished)> https://wandb.ai/edbeeching/trl/runs/eci6haqe
loading <Run edbeeching/trl/5nkq6xx5 (finished)> https://wandb.ai/edbeeching/trl/runs/5nkq6xx5
loading <Run edbeeching/trl/4lg964k3 (finished)> https://wandb.ai/edbeeching/trl/runs/4lg964k3
loading <Run edbeeching/trl/47q7jd52 (finished)> https://wandb.ai/edbeeching/trl/runs/47q7jd52
loading <Run edbeeching/trl/19jjjezh (finished)> https://wandb.ai/edbeeching/trl/runs/19jjjezh
loading <Run edbeeching/trl/uljkaufp (finished)> https://wandb.ai/edbeeching/trl/runs/uljkaufp
loading <Run edbeeching/trl/hnljrl4q (finished)> https://wandb.ai/edbeeching/trl/runs/hnljrl4q
loading <Run edbeeching/trl/smh5fo5m (finished)> https://wandb.ai/edbeeching/trl/runs/smh5fo5m
loading <Run edbeeching/trl/fze93818 (finished)> https://wandb.ai/edbeeching/trl/runs/fze93818
loading <Run edbeeching/trl/lvpnp3kj (finished)> https://wandb.ai/edbeeching/trl/runs/lvpnp3kj
loading <Run edbeeching/trl/npcqz80j (finished)> https://wandb.ai/edbeeching/trl/runs/npcqz80j
loading <Run edbeeching/trl/ft2l6sml (finished)> https://wandb.ai/edbeeching/trl/runs/ft2l6sml
loading <Run edbeeching/trl/9w45xxmp (finished)> https://wandb.ai/edbeeching/trl/runs/9w45xxmp
loading <Run edbeeching/trl/vlxw60fb (finished)> https://wandb.ai/edbeeching/trl/runs/vlxw60fb
loading <Run edbeeching/trl/stlnz79s (finished)> https://wandb.ai/edbeeching/trl/runs/stlnz79s
loading <Run edbeeching/trl/qckkkdwj (finished)> https://wandb.ai/edbeeching/trl/runs/qckkkdwj
loading <Run edbeeching/trl/lw5p22bi (finished)> https://wandb.ai/edbeeching/trl/runs/lw5p22bi
loading <Run edbeeching/trl/cqekbc0x (finished)> https://wandb.ai/edbeeching/trl/runs/cqekbc0x
loading <Run edbeeching/trl/3nnsujmm (finished)> https://wandb.ai/edbeeching/trl/runs/3nnsujmm
loading <Run edbeeching/trl/v0b38uab (finished)> https://wandb.ai/edbeeching/trl/runs/v0b38uab
loading <Run edbeeching/trl/o3tj6c4u (finished)> https://wandb.ai/edbeeching/trl/runs/o3tj6c4u
loading <Run edbeeching/trl/m4ubujss (crashed)> https://wandb.ai/edbeeching/trl/runs/m4ubujss
loading <Run edbeeching/trl/k2y9jz1o (crashed)> https://wandb.ai/edbeeching/trl/runs/k2y9jz1o
loading <Run edbeeching/trl/g0vc4wuu (crashed)> https://wandb.ai/edbeeching/trl/runs/g0vc4wuu
loading <Run edbeeching/trl/idttsf20 (crashed)> https://wandb.ai/edbeeching/trl/runs/idttsf20
loading <Run edbeeching/trl/g37nzx14 (crashed)> https://wandb.ai/edbeeching/trl/runs/g37nzx14
loading <Run edbeeching/trl/5wfs3l89 (crashed)> https://wandb.ai/edbeeching/trl/runs/5wfs3l89
kl-penalty-kl (sentiment analysis) has 10 runs
/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/openrlbenchmark/rlops_multi_metrics.py:388: RuntimeWarning: Mean of empty slice.
  result += [f"{metric_result.mean():.2f} ± {metric_result.std():.2f}"]
/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/numpy/core/_methods.py:206: RuntimeWarning: Degrees of freedom <= 0 for slice
  ret = _var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/numpy/core/_methods.py:163: RuntimeWarning: invalid value encountered in divide
  arrmean = um.true_divide(arrmean, div, out=arrmean,
/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/numpy/core/_methods.py:198: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
kl-penalty-mse (sentiment analysis) has 10 runs
kl-penalty-abs (sentiment analysis) has 10 runs
─────────────────────────────────────────────────────── Runtime (m) (mean ± std) ───────────────────────────────────────────────────────
┏━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Environment ┃ kl-penalty-kl (sentiment analysis) ┃ kl-penalty-mse (sentiment analysis) ┃ kl-penalty-abs (sentiment analysis) ┃
┡━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ trl         │ 18.514239360491434                 │ 18.54678015943159                   │ 18.13000280198298                   │
└─────────────┴────────────────────────────────────┴─────────────────────────────────────┴─────────────────────────────────────┘
───────────────────────────────────────────────────── Episodic Return (mean ± std) ─────────────────────────────────────────────────────
┏━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Environment ┃ kl-penalty-kl (sentiment analysis) ┃ kl-penalty-mse (sentiment analysis) ┃ kl-penalty-abs (sentiment analysis) ┃
┡━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ trl         │ nan ± nan                          │ nan ± nan                           │ nan ± nan                           │
└─────────────┴────────────────────────────────────┴─────────────────────────────────────┴─────────────────────────────────────┘
───────────────────────────────────────────────────────── Runtime (m) Average ──────────────────────────────────────────────────────────
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┓
┃ Environment                         ┃ Average Runtime    ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━┩
│ kl-penalty-kl (sentiment analysis)  │ 18.514239360491434 │
│ kl-penalty-mse (sentiment analysis) │ 18.54678015943159  │
│ kl-penalty-abs (sentiment analysis) │ 18.13000280198298  │
└─────────────────────────────────────┴────────────────────┘
saving figures and tables to static/0compare
saving report
Traceback (most recent call last):
  File "/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/fsx/ed/miniconda3/envs/trl_dev/lib/python3.10/site-packages/openrlbenchmark/rlops_multi_metrics.py", line 628, in <module>
    project=args.wandb_project_name,
AttributeError: 'Args' object has no attribute 'wandb_project_name'

Reply:
@vwxyzjn I will continue looking into this but I won't let it block the merging of this PR.